pkg/util/pki/parse.go

```diff
@@ -140,3 +140,161 @@ func DecodeX509CertificateRequestBytes(csrBytes []byte) (*x509.CertificateReques
 
 	return csr, nil
 }
+
+type PEMBundle struct {
+	CAPEM    []byte
+	ChainPEM []byte
+}
+
+type chainNode struct {
+	cert *x509.Certificate
+
+	issuer *chainNode
+}
+
+// ParseCertificateChainPEM will decode the bundle before passing to
+// ParseCertificateChain
+func ParseCertificateChainPEM(pembundle []byte) (*PEMBundle, error) {
+	certs, err := DecodeX509CertificateChainBytes(pembundle)
+	if err != nil {
+		return nil, err
+	}
+	return ParseCertificateChain(certs)
+}
+
+// ParseCertificateChain will return an x509 PEM encoded certificate chain and
+// CA. The CA may not be a true root, but the highest intermediate certificate.
+// Removes duplicate entries. The returned CA may be empty if a single
+// certificate was passed. Chain contains leaf certificate first. Removes
+// comments and white space. An error is returned if the passed bundle is not a
+// valid flat tree chain, the bundle is malformed, or the chain is broken.
+func ParseCertificateChain(certs []*x509.Certificate) (*PEMBundle, error) {
```

Comment:
I like returning a nil if there is an error. This syntax is clearer for consumers.

Reply:
The Go standard library does not makes use of pointers for indicating that the value is not present.

Instead, the presence of a non-nil error is used as a way to warn that the value itself can't be trusted

IMO this is a case where return a zero value makes sense, don't you think?

Reply:
Not sure what the intention here is with returning an empty struct rather than a pointer. What are we gaining from returning an empty struct?

Reply:
> What is the intention with returning an empty struct rather than a pointer

In this context, using the `nil` as a way to say "this function has errored" duplicates the purpose of the returned `error`. I would even go so far as to say that returning `nil` as a way to signify an error is a code smell since Go already has a mechanism for conveying errors.

IMO there are very few exceptions where a pointer makes more sense than returning a value:

1.  when the returned data contains something that is shared resource somewhere else (e.g., a struct that contains slices, or a struct that contains a mutex). Passing values (which implies copying) of shared resources does not make sense and would let the user think that this value can be copied.
2.  when a returned struct is big and that copying the value brings a performance penalty.

Since `PEMBundle` does not seem to contain any shared resource (since the returned \[\]byte are not shared anywhere else), and that the struct is tiny, it does not make sense to use a pointer to me.

Reply:
Don't think I agree with you here, but this is small thing so will just change it to value.

Reply:
I agree with Mael in this case, for what it's worth; I think the change to use a value is a good one.